data-prepper-expression/src/main/antlr/DataPrepperExpression.g4

```diff
@@ -0,0 +1,254 @@
+    ;
+
+regexEqualityOperator
+    : '=~'
+    | '!~'
+    ;
+
+relationalOperatorExpression
+    : relationalOperatorExpression relationalOperator setOperatorExpression
+    | setOperatorExpression
+    ;
+
+relationalOperator
+    : '<'
+    | '<='
+    | '>'
+    | '>='
+    ;
+
+setOperatorExpression
+    : setOperatorExpression setOperator setInitializer
+    | unaryOperatorExpression
+    ;
+
+setOperator
+    : 'in'
+    | 'not in'
+    ;
+
+unaryOperatorExpression
+    : primary
+    | setInitializer
+    | regexPattern
+    | parenthesisExpression
+    | unaryNotOperatorExpression
+    ;
+
+parenthesisExpression
+    : '(' conditionalExpression ')'
+    ;
+
+regexPattern
+    : jsonPointer
+    | String
+    ;
+
+setInitializer
+    : '{' primary (',' primary)* '}'
+    ;
+
+unaryNotOperatorExpression
+    : unaryOperator primary
+    ;
+
+unaryOperator
+    : 'not'
+    ;
+
+primary
+    : jsonPointer
+    | variableIdentifier
+    | setInitializer
+    | literal
+    ;
+
+jsonPointer
+    : JsonPointer
+    | EscapedJsonPointer
+    ;
+
+variableIdentifier
+    : variableName
+    ;
+
+variableName
+    : VariableIdentifier
+    ;
+
+literal
+    : Float
+    | Integer
+    | Boolean
+    | String
+    ;
+
+
+Integer
+    : '0'
+    | NonZeroDigit Digit*
+    ;
+
+Float
+    : NonZeroDigit? Digit '.' Digit
+    | NonZeroDigit? Digit '.' Digit+ NonZeroDigit
+    | '.' Digit
+    | '.' Digit* NonZeroDigit
+    ;
+
+fragment
+Digit
+    : '0'
+    | NonZeroDigit
+    ;
+
+fragment
+NonZeroDigit
+    : [1-9]
+    ;
+
+Boolean
+    : 'true'
+    | 'false'
+    ;
+
+JsonPointer
+    : FORWARDSLASH JsonPointerCharacters (FORWARDSLASH JsonPointerCharacters)*
+    ;
+
+fragment
+JsonPointerCharacters
+    : JsonPointerCharacter+
+    ;
+
+fragment
+JsonPointerCharacter
+    : [A-Za-z0-9_]
+    ;
+
+EscapedJsonPointer
+    : DOUBLEQUOTE FORWARDSLASH JsonPointerStringCharacters? DOUBLEQUOTE
+    ;
+
+fragment
+JsonPointerStringCharacters
+    : JsonPointerStringCharacter+
+    ;
+
+fragment
+JsonPointerStringCharacter
+    : ~["\\]
+    | JsonPointerEscapeSequence
+    ;
+
+fragment
+JsonPointerEscapeSequence
+    : '\\' [btnfr"'\\/]
+    ;
+
+VariableIdentifier
+    : '${' VariableNameCharacters '}'
+    ;
+
+fragment
+VariableNameCharacters
+    : VariableNameLeadingCharacter VariableNameCharacter*
+    ;
+
+fragment
+VariableNameLeadingCharacter
+    : [A-Za-z_]
+    ;
+
+fragment
+VariableNameCharacter
+    : VariableNameLeadingCharacter
+    | [0-9-]
+    ;
+
+String
+    : DOUBLEQUOTE StringCharacters? DOUBLEQUOTE
+    ;
+
+fragment
+StringCharacters
+    : StringCharacter+
+    ;
+
+fragment
+StringCharacter
+    : ~["\\$]
+    | EscapeSequence
+    ;
+
+fragment
+EscapeSequence
+    : '\\' [btnfr"'\\$]
+    ;
+
+EQUAL : '==';
+NOT_EQUAL : '!=';
+LT : '<';
+GT : '>';
+LTE : '<=';
+GTE : '>=';
+MATCH_REGEX_PATTERN : '=~';
+NOT_MATCH_REGEX_PATTERN : '!~';
+IN_SET : 'in';
+NOT_IN_SET : 'not in';
+AND : 'and';
```

Comment:
It appears that you have removed some of the literals from here. So I believe you can remove a few of these. I noticed `AND`, `OR`, LPAREN` and `RPAREN` at least can be removed.

Personally, I liked it when before, but I'm fine with either. But we should remove the unused literals from here to avoid possible confusion down the line.

Reply:
`LBRACK` and `RBRACK` removed.

Antlr uses the lexer rules (start with a lower case) to generate a list of symbol type int in Java. Here are the ones being generated -
```
	public static final int
		Integer=1, Float=2, Boolean=3, JsonPointer=4, EscapedJsonPointer=5, VariableIdentifier=6, 
		String=7, EQUAL=8, NOT_EQUAL=9, LT=10, GT=11, LTE=12, GTE=13, MATCH_REGEX_PATTERN=14, 
		NOT_MATCH_REGEX_PATTERN=15, IN_SET=16, NOT_IN_SET=17, AND=18, OR=19, NOT=20, 
		LPAREN=21, RPAREN=22, LBRACE=23, RBRACE=24, FORWARDSLASH=25, DOUBLEQUOTE=26, 
		SET_SEPARATOR=27, PERIOD=28, SPACE=29, OTHER=30;
```

Parser rules start with a lower case letter. We can define parser rules with exact string matches or lexer rules directly. Here is what either might look like
```
Boolean : 'true' | 'false';

conditionalOperator : 'true' | 'false';
// or
conditionalOperator : Boolean;
```

Both function exactly the same, it a preference on grammar style. If the parser rule `conditionalOperator` the TerminalNode will be a symbol of type `Boolean`. Using the symbol type on the Java side will look like this:

```
((TerminalNode) node).getSymbol().getType() == DataPrepperExpressionParser.Boolean
```

In a revision of this PR I removed the lexer rules for `TRUE : 'true'; FALSE : 'false';` to keep boolean literal consistent with other literals. Now all literal can be checked the same way:
```
((TerminalNode) node).getSymbol().getType() == DataPrepperExpressionParser.Integer;
((TerminalNode) node).getSymbol().getType() == DataPrepperExpressionParser.Float;
((TerminalNode) node).getSymbol().getType() == DataPrepperExpressionParser.Boolean;
((TerminalNode) node).getSymbol().getType() == DataPrepperExpressionParser.String;
```

Before that revision boolean would need to be checked like this
```
int symbolType = ((TerminalNode) node).getSymbol().getType();
symbolType == DataPrepperExpressionParser.TRUE || symbolType == DataPrepperExpressionParser.FALSE
```